File: ch4/akkademy-db-java/src/main/java/com/akkademy/AkkademyDb.java
package com.akkademy;

import akka.actor.AbstractActor;
import akka.actor.Status;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.pf.ReceiveBuilder;
import com.akkademy.messages.Connected;
import com.akkademy.messages.GetRequest;
import com.akkademy.messages.KeyNotFoundException;
import com.akkademy.messages.SetRequest;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class AkkademyDb extends AbstractActor {
    protected final LoggingAdapter log = Logging.getLogger(context().system(), this);
    public final Map<String, Object> map = new HashMap<String, Object>();

    private AkkademyDb() {
        receive(ReceiveBuilder.
                        match(Connected.class, message -> {
                            log.info("Received Connected request: {}", message);
                            sender().tell(new Connected(), self());
                        }).
                        match(List.class, message -> {
                                    message.forEach(x -> {
                                        if (x instanceof SetRequest) {
                                            SetRequest setRequest = (SetRequest) x;
                                            handleSetRequest(setRequest);
                                        }
                                        if (x instanceof GetRequest) {
                                            GetRequest getRequest = (GetRequest) x;
                                            handleGetRequest(getRequest);
                                        }
                                    });
                                }
                        ).
                        match(SetRequest.class, message -> {
                            handleSetRequest(message);
                        }).
                        match(GetRequest.class, message -> {
                            handleGetRequest(message);
                        }).
                        matchAny(o -> {
                            log.info("unknown message: " + o);
                            sender().tell(new Status.Failure(new ClassNotFoundException()), self());
                        }).build()
        );
    }

    private void handleSetRequest(SetRequest message) {
        log.info("Received Set request: {}", message);
        map.put(message.key, message.value);
        message.sender.tell(new Status.Success(message.key), self());
    }

    private void handleGetRequest(GetRequest getRequest) {
        log.info("Received Get request: {}", getRequest);
        Object value = map.get(getRequest.key);
        Object response = (value != null)
                ? value
                : new Status.Failure(new KeyNotFoundException(getRequest.key));
        sender().tell(response, self());
    }
}

File: ch4/akkademy-db-scala/src/main/scala/com/akkademy/AkkademyDb.scala
package com.akkademy

import akka.actor._
import akka.event.Logging
import com.akkademy.messages.{KeyNotFoundException, GetRequest, SetRequest}
import scala.collection.mutable.HashMap

class AkkademyDb extends Actor {
  val map = new HashMap[String, Object]
  val log = Logging(context.system, this)

  override def receive = {
    case x: messages.Connected =>
      sender() ! x
    case x: List[_] =>
      x.foreach{
        case SetRequest(key, value, senderRef) =>
          handleSetRequest(key, value, senderRef)
        case GetRequest(key, senderRef) =>
          handleGetRequest(key, senderRef)
      }
    case SetRequest(key, value, senderRef) =>
      handleSetRequest(key, value, senderRef)
    case GetRequest(key, senderRef) =>
      handleGetRequest(key, senderRef)
    case o =>
      log.info("unknown message")
      sender() ! Status.Failure(new ClassNotFoundException)
  }

  def handleSetRequest(key: String, value: Object, senderRef: ActorRef): Unit = {
    log.info("received SetRequest - key: {} value: {}", key, value)
    map.put(key, value)
    senderRef ! Status.Success
  }

  def handleGetRequest(key: String, senderRef: ActorRef): Unit = {
    log.info("received GetRequest - key: {}", key)
    val response: Option[Object] = map.get(key)
    response match {
      case Some(x) => senderRef ! x
      case None => senderRef ! Status.Failure(new KeyNotFoundException(key))
    }
  }
}

object Main extends App {
  val system = ActorSystem("akkademy")
  val helloActor = system.actorOf(Props[AkkademyDb], name = "akkademy-db")
}
